src/p_spec.c

```diff
@@ -179,6 +207,43 @@ void P_InitPicAnims (void)
   Z_ChangeTag (animdefs,PU_CACHE); //jff 3/23/98 allow table to be freed
 }
 
+// [FG] play sound when hitting animated floor
+void P_HitFloor (mobj_t *mo)
+{
+  const sector_t *sec = mo->subsector->sector;
+  const short floorpic = sec->floorpic;
+  terrain_t terrain = terrain_solid;
+  anim_t *anim;
+
+  if (mo->floorz != sec->floorheight)
+  {
+    return;
+  }
+
+  for (anim = anims ; anim < lastanim ; anim++)
```

Comment:
What do you think about just making it a sector property so you can do `sec->terrain` without doing a search? You would have to set sector terrain anytime the floorpic is set.

Reply:
Yes, I already had this in mind, but there are several places where the floor pic could change,  e.g. donuts etc. 🤔

Reply:
Sure, but even then I would have to iterate through the anim list to see if the new floorpic falls into one of the animation ranges.

Reply:
For mid-level changes you could actually just do `sec->terrain = other_sec->terrain` when copying one sector's floorpic to the other. So the abstraction would be at the sector property copying instead of floorpic specifically.

Reply:
Hm actually it looks like it's not always that simple in dsda-doom at least. Sometimes the thinker carries the texture with it so not every copy has the sector in context.

Reply:
Or, I could allocate an array of size `numflats` as a lookup table and set the terrain for each flat there.